src/test/java/com/commercetools/sync/services/impl/UnresolvedReferencesServiceImplTest.java

```diff
@@ -0,0 +1,197 @@
+                        valueObj.getProductDraft().getKey(),
+                        valueObj,
+                        WaitingToBeResolved.class);
+        verify(productSyncOptions.getCtpClient()).execute(eq(CustomObjectUpsertCommand.of(customObjectDraft)));
+    }
+
+    @Test
+    void save_WithUnsuccessfulMockCtpResponse_ShouldNotSaveMock() {
+        final CustomObject mock = mock(CustomObject.class);
+        when(mock.getContainer()).thenReturn("commercetools-sync-java.UnresolvedReferencesService.productDrafts");
+
+        final ProductDraft productDraft = mock(ProductDraft.class);
+        when(productDraft.getKey()).thenReturn("product-draft-key");
+        final Set<String> missingRefs = new HashSet<>();
+        missingRefs.add("test-ref");
+        WaitingToBeResolved valueObj = new WaitingToBeResolved(productDraft, missingRefs);
+        when(mock.getValue()).thenReturn(valueObj);
+
+        when(productSyncOptions.getCtpClient().execute(any()))
+                .thenReturn(CompletableFutureUtils.failed(new BadRequestException("bad request")));
+
+        final Optional<WaitingToBeResolved> toBeResolvedOptional = service.save(valueObj).toCompletableFuture().join();
+
+        assertThat(toBeResolvedOptional).isEmpty();
+        assertThat(errorMessages)
+                .hasSize(1)
+                .hasOnlyOneElementSatisfying(message -> {
+                    assertThat(message)
+                            .contains("Failed to save CustomObject with key: 'product-draft-key'.");
+                    assertThat(message).contains("BadRequestException");
+                });
+        assertThat(errorExceptions)
+                .hasSize(1)
+                .hasOnlyOneElementSatisfying(exception ->
+                        assertThat(exception).isExactlyInstanceOf(BadRequestException.class));
+    }
+
+
+    @Test
+    void save_WithDraftWithoutKey_ShouldNotSaveWaitingToBeResolved() {
+        final WaitingToBeResolved draft = mock(WaitingToBeResolved.class);
+        when(draft.getProductDraft()).thenReturn(mock(ProductDraft.class));
+        final Optional<WaitingToBeResolved> beResolvedOptional = service
+                .save(draft).toCompletableFuture().join();
+
+        assertThat(beResolvedOptional).isEmpty();
+        assertThat(errorMessages).hasSize(1);
+        assertThat(errorExceptions).hasSize(1);
+        assertThat(errorMessages.get(0))
+                .isEqualTo("Failed to save CustomObject with key: 'null'. Reason: Draft key is blank!");
+    }
+
+    @Test
+    void fetch_WithEmptyKeySet_ShouldReturnEmptySet() {
+        Set<String> keys = new HashSet<>();
+        keys.add("");
+        final Set<WaitingToBeResolved> beResolvedOptional = service
+                .fetch(keys).toCompletableFuture().join();
+
+        assertThat(beResolvedOptional).isEmpty();
+    }
+
+
+    @Test
+    void fetch_WithSuccessfulMockCtpResponse_ShouldReturnMock() {
```

Comment:
what about fetch with unsuccessful?

Reply:
I find tests to read when the bodies are separated with comments:

```java
// preparation
...
// test
...
// assertion
...
```
What do you think about this?

Reply:
> what about fetch with unsuccessful?

As we are calling this static method `QueryExecutionUtils.queryAll` and it does not possible to mock the handle the failed called directly.

Reply:
cc8958914da78b9df1fa42bb9ae1cf36db93b815